### PR TITLE
GCW-3388 Add restrictions for mentioning users

### PR DESCRIPTION
**Pull request by bharat619**
### Ticket Link: 
https://jira.crossroads.org.hk/browse/GCW-3388
NOTE: Specify ticket link

### CODE REVIEW

File: app/models/concerns/mentionable.rb [hunk 1]
```diff
@@ -48,20 +48,27 @@ def parse_id_from_decorated_ids(ids)
...
 
   included do
-    def self.mentionable_users(opts = {})
-      users = active.exclude_user(User.current_user.id)
-                    .with_roles(mentionable_roles(opts[:roles])).distinct.to_a
-      users << add_order_creator(users, opts[:order_id]) if opts[:order_id]
+    def self.mentionable_users(roles:, order_id:, is_private: false)
+      users = User.active.exclude_user(User.current_user.id)
+                  .with_roles(mentionable_roles(roles)).distinct.to_a
+      users << add_order_creator(users, order_id) unless bool_cast.call(is_private)
```
Comment (patrixr) [inline app/models/concerns/mentionable.rb:54]: Looks good, but I would make it generic to any resource type, not just order. 
For sure in the coming months we'll talk about mentionable users on offers, etc

File: app/models/concerns/mentionable.rb [hunk 2]
```diff
@@ -48,20 +48,27 @@ def parse_id_from_decorated_ids(ids)
...
       user = Order.find_by(id: order_id)&.created_by
       user unless users.include? user
     end
 
     def self.mentionable_roles(roles)
       (MENTIONABLE_ROLES & roles.split(',').map(&:strip).uniq)
     end
+
+    def self.bool_cast
+      ->(x) { ActiveModel::Type::Boolean.new.cast(x) }
```
Comment (patrixr) [inline app/models/concerns/mentionable.rb:71]: why the proc ?

```suggestion
      ActiveModel::Type::Boolean.new.cast(x)
```

File: app/controllers/api/v1/users_controller.rb [hunk 3]
```diff
@@ -76,8 +76,9 @@ def orders_count
       def mentionable_users
         return render json: { users: [] } if params['roles'].nil?
 
-        @users = User.mentionable_users({ roles: params[:roles],
-                                          order_id: params[:order_id] })
+        @users = User.mentionable_users( roles: params[:roles],
+                                         order_id: params[:order_id],
```
Comment (patrixr) [inline app/controllers/api/v1/users_controller.rb:80]: Same as the comment below, messages are not order-specific, so if possible this shouldn't either

File: app/models/concerns/mentionable.rb [hunk 4]
```diff
@@ -48,20 +48,27 @@ def parse_id_from_decorated_ids(ids)
...
-      users << add_order_creator(users, opts[:order_id]) if opts[:order_id]
+    def self.mentionable_users(roles:, order_id:, is_private: false)
+      users = User.active.exclude_user(User.current_user.id)
+                  .with_roles(mentionable_roles(roles)).distinct.to_a
+      users << add_order_creator(users, order_id) unless bool_cast.call(is_private)
       users.compact
     end
 
     def self.add_order_creator(users, order_id)
+      valid_roles = User.current_user.roles.pluck(:name) & ['Order administrator', 'Order fulfilment']
```
Comment (steveyken) [inline app/models/concerns/mentionable.rb:59]: `add_order_creator` is very specific to `order`. Can you use `mentionable_roles` or a different approach?

**Review by patrixr — APPROVED** [review]

File: app/models/concerns/mentionable.rb [hunk 5]
```diff
@@ -48,27 +55,37 @@ def parse_id_from_decorated_ids(ids)
   end
 
   included do
-    def self.mentionable_users(roles:, order_id:, is_private: false)
+    def self.mentionable_users(roles:, messageable_id:, messageable_type:, is_private: false)
+      messageable = construct_messageable(messageable_type, messageable_id) unless [messageable_id, messageable_type].all?(&:nil?)
       users = User.active.exclude_user(User.current_user.id)
                   .with_roles(mentionable_roles(roles)).distinct.to_a
```
Comment (steveyken) [inline app/models/concerns/mentionable.rb:61]: This should probably include only users who have roles that are not expired.

**Review by steveyken — APPROVED** [review]
Thanks for the changes, Much better. One final thing I spotted.

**Pull request merged**